File: ui/src/api/phase3.ts
export type StoragePool = {
  id: string;
  name: string;
  backend_type: string;
  status: string;
  reason?: string;
  locator?: string;
  warnings?: string[];
  warning_text?: string[];
  capabilities?: {
    incremental_send?: boolean;
    snapshots?: boolean;
    volume_create?: boolean;
    sparse_files?: boolean;
    xattr_identity?: boolean;
    shared_warning?: boolean;
    supported_classes?: string[];
  };
  usable_bytes?: number | null;
  allocated_bytes?: number | null;
  provisioned_bytes?: number | null;
  total_bytes?: number | null;
  metadata_percent?: number | null;
  storage_classes?: string[];
  adopted?: boolean;
};

export type StorageVolume = {
  id: string;
  pool_id: string;
  class: string;
  kind: string;
  format: string;
  size_bytes: number;
  status: string;
  backend_type: string;
  backend_ref: string;
  xattr_state?: string;
  allocated_bytes?: number | null;
};

export type LibraryItem = {
  id: string;
  pool_id: string;
  kind: string;
  display_name: string;
  backend_ref: string;
  size_bytes: number;
  checksum_sha256: string;
  status: string;
  created_at?: string;
};

export type PoolListResponse = {
  items: StoragePool[];
  default_path?: string;
};

export type ZFSRuntime = {
  backend?: string;
  incremental_send?: boolean;
  snapshots?: boolean;
  directory_default?: boolean;
  force_import?: string;
  host_supported?: boolean;
  status?: string;
  reason?: string;
  packages?: string[];
};

export type LVMRuntime = {
  backend?: string;
  incremental_send?: boolean;
  snapshots?: boolean;
  directory_default?: boolean;
  vgexport?: string;
  host_supported?: boolean;
  status?: string;
  reason?: string;
  packages?: string[];
};

export type DatastoreRuntime = {
  nfs?: boolean;
  smb?: boolean;
  iscsi?: boolean;
  incremental_send?: boolean;
  directory_default?: boolean;
  passwords_in_unit_files?: boolean;
  host_supported?: boolean;
  status?: string;
  reason?: string;
  packages?: string[];
};

export type DistributedRuntime = {
  backend?: string;
  incremental_send?: boolean;
  directory_default?: boolean;
  keys_in_list_json?: boolean;
  feature_enabled?: boolean;
  osd_process?: boolean;
  osd_started?: boolean;
  host_supported?: boolean;
  status?: string;
  reason?: string;
  vm_disk_rbd?: boolean;
};

File: internal/appdb/phase3.go
package appdb

import (
	"encoding/json"
	"time"

	"github.com/no-dal/ndl-ce/internal/storage"
)

// StoragePool is desired existence plus observed availability.
// RootPath and Backing are locators, not identity. ID is the UUID.
type StoragePool struct {
	ID               string
	ClusterID        string
	NodeID           string
	Name             string
	BackendType      string
	Status           string
	Reason           string
	RootPath         string
	Backing          json.RawMessage
	Warnings         []string
	WarningText      []string
	Capabilities     json.RawMessage
	UsableBytes      *int64
	AllocatedBytes   *int64
	ProvisionedBytes *int64
	TotalBytes       *int64
	Adopted          bool
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

// Volume is identified by UUID. BackendRef is a locator.
type Volume struct {
	ID             string
	ClusterID      string
	NodeID         string
	PoolID         string
	Class          string
	Kind           string
	Format         string
	SizeBytes      int64
	Status         string
	BackendType    string
	BackendRef     string
	XattrState     string
	AllocatedBytes *int64
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// LibraryItem is identified by UUID. DisplayName is metadata.
type LibraryItem struct {
	ID             string
	ClusterID      string
	NodeID         string
	PoolID         string
	Kind           string
	DisplayName    string
	BackendRef     string
	SizeBytes      int64
	ChecksumSHA256 string
	Status         string
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

func poolHint(p StoragePool) storage.PoolHint {
	var backing storage.BackingIdentity
	if len(p.Backing) > 0 {
		_ = json.Unmarshal(p.Backing, &backing)
	}
	return storage.PoolHint{
		PoolID:      p.ID,
		BackendType: p.BackendType,
		RootPath:    p.RootPath,
		Backing:     backing,
	}
}

func PoolHints(pools []StoragePool) []storage.PoolHint {
	out := make([]storage.PoolHint, 0, len(pools))
	for _, p := range pools {
		out = append(out, poolHint(p))
	}
	return out
}
